### PR TITLE
Fix PATH priority for windows

### LINKED ISSUE #6175: Bug: Postinstall Scripts of Workspace Packages are called within the node_modules of the workspace
This bug comes up when trying to develop a vscode extension using yarn's workspace feature.
# Setup
Yarn v1.9.2, Windows 10, Node v10.6.0
Directory layout:
* ROOT
    * package.json
    ```json
    {
        "private": true,
        "workspaces": [ "vscode-ext", "common"]
    }
    ```
    * vscode-ext
      * package.json
    ```json
    ...
        "name": "vscode-ext",
        "scripts": {
            "postinstall": "yarn run vscode-install"
        },
    "devDependencies": { "vscode": "^1.1.6" }
    ...
    ```
    * common (unimportant)

Yarn install turns it into this structure:
* ROOT
    * node_modules
      * vscode-ext (symlink to ROOT/vscode-ext)
      * vscode
        * .bin
          * install
    * package.json
    * vscode-ext
      * package.json
      * node_modules
        * bin
          * vscode-install.cmd
    ```batch
    @IF EXIST "%~dp0\node.exe" (
    "%~dp0\node.exe"  "%~dp0\..\..\..\node_modules\vscode\bin\install" %*
    ) ELSE (
    @SETLOCAL
    @SET PATHEXT=%PATHEXT:;.JS;=;%
    node  "%~dp0\..\..\..\node_modules\vscode\bin\install" %*
    )
    ```
    * common (unimportant)

# Problem
The postinstall script of the vscode-ext package is run within `ROOT/node_modules/vscode-ext` and not within `ROOT/vscode-ext`.
Both paths point to the same directory, but relativ paths behave differently:

In `ROOT/node_modules/vscode-ext/node_modules/.bin/`,
`"%~dp0\..\..\..\node_modules\vscode\bin\install" %*` points to `ROOT/node_modules/node_modules/vscode/bin/install` which is clearly wrong.

In `ROOT/vscode-ext/node_modules/.bin/` however,
`"%~dp0\..\..\..\node_modules\vscode\bin\install" %*` points to `ROOT/node_modules/vscode/bin/install` which actually works.

This behavior breaks at least vscode extensions in yarn workspaces.

### PR DESCRIPTION
**Pull request by 3846masa**
**Summary**

> Fix #6175 

The example code caused this issue is [here](https://github.com/3846masa-tmp/issue-6175-yarn).

---

On windows, yarn create `.cmd` file as below when installing workspaces.

```cmd
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\..\..\..\node_modules\rimraf\bin.js" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\..\..\..\node_modules\rimraf\bin.js" %*
)
```

However, `%~dp0` cannot resolve symlink path, so `%~dp0` maybe changed.

---

For example,  you have workspaces as below,

```
root
|- package.json
+- package-a
   +- package.json
```

`package-a/package.json` will install `rimraf`.

After running `yarn install`, change to below,

```
root
|- package.json
|- node_modules
|  |- rimraf
|  |  +- bin.js
|  +- .bin
|  |  +- rimraf.cmd
|  +- package-a // symlink to root/package-a
|     +- node_modules
|        +- .bin
|           +- rimraf.cmd
+- package-a
   |- package.json
   +- node_modules
      +- .bin
         +- rimraf.cmd
```

`package-a/node_modules/.bin/rimraf.cmd` and `node_modules/package-a/node_modules/.bin/rimraf.cmd` are same file as below,

```cmd
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\..\..\..\node_modules\rimraf\bin.js" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\..\..\..\node_modules\rimraf\bin.js" %*
)
```

`%~dp0\..\..\..\node_modules\rimraf\bin.js` from `package-a/node_modules/.bin/rimraf.cmd` is `node_modules/rimraf/bin.js`.

However, after installing, postinstall script will be resolved from installed folder.
The installed folder in this example is `node_modules/package-a`.

Because of that, postinstall run `node_modules/pacakge-a/node_modules/.bin/rimraf.cmd`.
`%~dp0\..\..\..\node_modules\rimraf\bin.js` from  `node_modules/pacakge-a/node_modules/.bin/rimraf.cmd` is `node_modules/node_modules/rimraf/bin.js`.
This is invalid path.

---

This PR fixes PATH priority for using `node_modules/.bin` when workspace mode.
In other words, after installing, postinstall script will be resolved from root folder at first.

**Test plan**

Run `yarn install` in root folder on Windows using [this sample code](https://github.com/3846masa-tmp/issue-6175-yarn).

### CODE REVIEW

**Comment by ninja-** [comment]
interesting. you saw my alternate fix btw? https://github.com/yarnpkg/yarn/pull/7299

**Comment by 3846masa** [comment]
@ninja- 
I tried your PR, however it cannot work on my environment (Windows 10 / MSYS 2).

**Comment by ninja-** [comment]
I am sure it fixed the issue for me. What's the result if you run from cmd instead?

**Comment by 3846masa** [comment]
I tried your PR with [this code](https://github.com/3846masa-tmp/issue-6175-yarn) on cmd and powershell, but it cannot work. (Windows 10 / Node.js v10.15.1)
**Please talk about your PR in your PR.**

**Comment by ninja-** [comment]
I checked on my project and it's not fixing the issue. I still have the same error on my project. Maybe we should pull both of our patches to fix the edge cases.

**Comment by ninja-** [comment]
@arcanis

**Comment by ninja-** [comment]
> This PR fixes PATH priority for using node_modules/.bin when workspace mode.

for some reason your patch is not doing that, both on cygwin and cmd

**Comment by ninja-** [comment]
I did some digging and it looks like when compiling submodules of submodules workspaceEnabled is not getting trigerred and the workspace .bin is not being pushed into array...in simpler cases it's fine.

main depends on somelib
somelib depends on common

all of these are workspace modules

then the paths are

[ 'C:\\Users\\x\\IdeaProjects\\workspace\\node_modules\\.bin',
  'C:\\Users\\x\\IdeaProjects\\workspace\\node_modules\\somelib\\node_modules\\.bin',

in other cases it would be correctly

[ 'C:\\Users\\x\\IdeaProjects\\workspace\\node_modules\\.bin',
  'C:\\Users\\x\\IdeaProjects\\workspace\\mai\\node_modules\\somelib\\node_modules\\.bin',
  'C:\\Users\\x\\AppData\\Local\\Yarn\\Data\\link\\node_modules\\.bin',

**Comment by 3846masa** [comment]
@ninja- For understanding your test case, please share your test case code via GitHub repository or any way?

**Comment by ninja-** [comment]
@3846masa I kind of gave up on it, and replaced yarn run with direct command. The goal was to don't-repeat-yourself a little so the package.json was like

```
"build": "tsc && do something && bla",
"install": "yarn run build"
```

now
```
"build": "tsc && do something && bla",
"install": "tsc && do something && bla"
```

so that fixes it, but would be nice to fix yarn run as well, but I don't know how.

As for your patch, I would like it to get merged ASAP so I don't need to ship patched yarn to coworkers working on the same project...

**Comment by ninja-** [comment]
@arcanis any chance for merge?

**Comment by greenreign** [comment]
Would love to have a fix for this.  Any chance someone can look at this PR?

**Comment by ninja-** [comment]
🙄
